keel-bakery-plugin/src/main/kotlin/com/netflix/spinnaker/keel/bakery/artifact/ImageHandler.kt

```diff
@@ -32,66 +31,63 @@ class ImageHandler(
-          } else {
-            log.warn("Detected a diff in the regions for ${current.appVersion}: ${diff.toDebug()}")
-            publisher.publishEvent(ImageRegionMismatchDetected(current, artifact.vmOptions.regions))
+        val desiredBaseAmiName = artifact.findLatestBaseAmiName()
+
+        val images = artifact.findLatestAmi(desiredAppVersion)
+        val olderBaseImages = images.filterValues { it.baseImageName != desiredBaseAmiName }
+        val missingRegions = artifact.vmOptions.regions - images.keys
+        when {
+          images.isEmpty() -> {
+            log.info("No AMI found for {}", desiredAppVersion)
+            launchBake(artifact, desiredAppVersion)
           }
-        } else if (diff.hasChanges()) {
-          if (current == null) {
-            log.info("No AMI found for {}", artifact.name)
-          } else {
-            log.info("Image for {} delta: {}", artifact.name, diff.toDebug())
+          olderBaseImages.isNotEmpty() -> {
+            log.info("AMIs for {} are outdated, rebaking…", desiredAppVersion)
+            launchBake(
+              artifact,
+              desiredAppVersion,
+              description = "Bake $desiredAppVersion due to a new base image: $desiredBaseAmiName"
+            )
           }
-
-          if (diffFingerprintRepository.seen("ami:${artifact.name}", diff)) {
-            log.warn("Artifact version {} and base AMI version {} were baked previously", latestArtifactVersion, latestBaseAmiName)
-            publisher.publishEvent(RecurrentBakeDetected(latestArtifactVersion, latestBaseAmiName))
-          } else {
-            launchBake(artifact, latestArtifactVersion, diff)
-            diffFingerprintRepository.store("ami:${artifact.name}", diff)
+          missingRegions.isNotEmpty() -> {
+            log.warn("Detected missing regions for ${desiredAppVersion}: ${missingRegions.joinToString()}")
+            publisher.publishEvent(ImageRegionMismatchDetected(desiredAppVersion, desiredBaseAmiName, images.keys, artifact.vmOptions.regions))
+            launchBake(
+              artifact,
+              desiredAppVersion,
+              regions = missingRegions,
+              description = "Bake $desiredAppVersion due to missing regions: ${missingRegions.joinToString()}"
+            )
+          }
+          else -> {
+            log.debug("Image for {} already exists with app version {} and base image {} in regions {}", artifact.name, desiredAppVersion, desiredBaseAmiName, artifact.vmOptions.regions.joinToString())
           }
-        } else {
-          log.debug("Existing image for {} is up-to-date", artifact.name)
         }
       }
     }
   }
 
-  private suspend fun DebianArtifact.findLatestAmi() =
-    imageService.getLatestImage(this, "test")
+  private suspend fun DebianArtifact.findLatestAmi(desiredArtifactVersion: String) =
+    imageService.getLatestNamedImages(AppVersion.parseName(desiredArtifactVersion), "test", vmOptions.regions)
```

Comment:
Now just using the same method that `ImageExistsConstraint` and `ImageResolver` use to get the AMIs for a specific app version